vhost-device-spi/src/vhu_spi.rs

```diff
@@ -0,0 +1,1206 @@
+// vhost device spi
+//
+// Copyright (c) 2024 Qualcomm Innovation Center, Inc. All rights reserved.
+//          Haixu Cui <quic_haixcui@quicinc.com>
+//
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+use std::{
+    cmp::Ordering,
+    convert,
+    io::{self, Read, Result as IoResult, Write},
+    mem::size_of,
+    slice::from_raw_parts,
+    sync::Arc,
+};
+
+use log::warn;
+use thiserror::Error as ThisError;
+use vhost::vhost_user::message::{VhostUserProtocolFeatures, VhostUserVirtioFeatures};
+use vhost_user_backend::{VhostUserBackendMut, VringRwLock, VringT};
+use virtio_bindings::bindings::virtio_config::{VIRTIO_F_NOTIFY_ON_EMPTY, VIRTIO_F_VERSION_1};
+use virtio_bindings::bindings::virtio_ring::{
+    VIRTIO_RING_F_EVENT_IDX, VIRTIO_RING_F_INDIRECT_DESC,
+};
+use virtio_queue::DescriptorChain;
+use virtio_queue::QueueOwnedT;
+use vm_memory::{
+    ByteValued, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryLoadGuard, GuestMemoryMmap, Le32,
+};
+use vmm_sys_util::epoll::EventSet;
+use vmm_sys_util::eventfd::{EventFd, EFD_NONBLOCK};
+
+use crate::spi::*;
+
+const QUEUE_SIZE: usize = 1024;
+const NUM_QUEUES: usize = 1;
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, ThisError)]
+/// Errors related to vhost-device-spi daemon.
+pub(crate) enum Error {
+    #[error("TX length  {0} and RX length {1} don't match")]
+    TxRxTrnasLenNotEqual(u32, u32),
+    #[error("TX length and RX length are both zero")]
```

Comment:
I'd suggest more descriptive error messages, like "TX length and RX length in SPI request do not match" and also add a debug print of the request itself, perhaps.

Reply:
are you saying that I should remove the two arguments of `TxRxTrnasLenNotEqual`, and add a `warn!("invalid request information");` before return `Error::TxRxTrnasLenNotEqual`.